datacube/drivers/postgres/_api.py

```diff
@@ -535,6 +536,77 @@ def search_datasets(self, expressions,
                                                   select_fields, with_source_ids, limit)
         return self._connection.execute(select_query)
 
+    @staticmethod
+    def search_unique_datasets_query(expressions, select_fields, limit):
+        """
+        We are not dealing with dataset_source table here and we are not joining
+        dataset table with dataset_location table. We are aggregating stuff
+        in dataset_location per dataset basis if required. It returns the construted
+        query.
+        """
+
+        # expressions involving DATASET_SOURCE cannot not done for now
+        for expression in expressions:
+            assert expression.field.required_alchemy_table != DATASET_SOURCE,\
+                'Joins with dataset_source cannot be done for this query'
+
+        # expressions involving 'uri' and 'uris' will be handled different
+        expressions = [expression for expression in expressions
+                       if expression.field.required_alchemy_table != DATASET_LOCATION]
+
+        if select_fields:
+            select_columns = []
+            for field in select_fields:
+                if field.name in {'uri', 'uris'}:
+                    # All active URIs, from newest to oldest
+                    uris_field = func.array(
+                        select([
+                            _dataset_uri_field(SELECTED_DATASET_LOCATION)
+                        ]).where(
+                            and_(
+                                SELECTED_DATASET_LOCATION.c.dataset_ref == DATASET.c.id,
+                                SELECTED_DATASET_LOCATION.c.archived == None
+                            )
+                        ).order_by(
+                            SELECTED_DATASET_LOCATION.c.added.desc(),
+                            SELECTED_DATASET_LOCATION.c.id.desc()
+                        ).label('uris')
+                    ).label('uris')
+                    select_columns.append(uris_field)
+                else:
+                    select_columns.append(field.alchemy_expression.label(field.name))
+        else:
+            select_columns = _DATASET_SELECT_FIELDS
+
+        raw_expressions = PostgresDbAPI._alchemify_expressions(expressions)
+
+        # We don't need 'DATASET_LOCATION table in the from expression
+        select_fields_ = [field for field in select_fields if field.name not in {'uri', 'uris'}]
+
+        from_expression = PostgresDbAPI._from_expression(DATASET, expressions, select_fields_)
+        where_expr = and_(DATASET.c.archived == None, *raw_expressions)
+
+        return(
+            select(
+                select_columns
+            ).select_from(
+                from_expression
+            ).where(
+                where_expr
+            ).limit(
+                limit
+            )
+        )
+
+    def search_unique_datasets(self, expressions, select_fields=None, limit=None):
+        """
+        Processes a search query without duplicating datasets.
```

Comment:
needs better doc string, what does "unique" mean,  how is it different from just `search_datasets"

Reply:
Done